tests/S3/S3Transfer/S3TransferManagerTest.php

```diff
@@ -0,0 +1,3869 @@
+                'expectations' => $case['expectations'],
+                'outcomes' => $case['outcomes'],
+            ];
+        }
+    }
+
+    /**
+     * @return Generator
+     */
+    public function modeledDownloadDirectoryCasesProvider(): Generator
+    {
+        $downloadCases = json_decode(
+            file_get_contents(
+                self::DOWNLOAD_DIRECTORY_BASE_CASES
+            ),
+            true
+        );
+        foreach ($downloadCases as $case) {
+            yield $case['summary'] => [
+                'test_id' => $case['summary'],
+                'config' => $case['config'],
+                'download_directory_request' => $case['downloadDirectoryRequest'],
+                's3_objects' => $case['s3Objects'],
+                'expectations' => $case['expectations'],
+                'expected_files' => $case['expectedFiles'],
+                'outcomes' => $case['outcomes'],
+            ];
+        }
+    }
+
+    /**
+     * @param array $requestArgs
+     *
+     * @return void
+     */
+    private function parseRequestArgsFromCamelCaseToPascalCase(
+        array &$requestArgs
+    ): void
+    {
+        foreach ($requestArgs as $key => $value) {
+            $newKey = ucfirst($key);
+            unset($requestArgs[$key]);
+            $requestArgs[$newKey] = $value;
+        }
+    }
+
+    /**
+     * @param array $caseHeaders
+     *
+     * @return void
+     */
+    private function parseCaseHeadersToAmzHeaders(array &$caseHeaders): void
+    {
+        foreach ($caseHeaders as $key => $value) {
+            $newKey = $key;
+            switch ($key) {
+                case 'PartsCount':
+                    $newKey = 'x-amz-mp-parts-count';
+                    break;
+                case 'ChecksumAlgorithm':
+                    $newKey = 'x-amz-checksum-algorithm';
+                    break;
+                default:
+                    if (preg_match('/Checksum[A-Z]+/', $key)) {
+                        $newKey = 'x-amz-checksum-' . str_replace(
+                            'Checksum',
+                            '',
+                            $key
+                        );
+                    }
+            }
+
+            if ($newKey !== $key) {
+                $caseHeaders[$newKey] = $value;
+                unset($caseHeaders[$key]);
+            }
+        }
+    }
+
+    /**
+     * @param array $methodsCallback If any from the callbacks below
+     *  is not provided then a default implementation will be provided.
+     * - getCommand: (Closure, optional) This callable will
+     *   receive as parameters:
+     *   - $commandName: (string, optional)
+     *   - $args: (array, optional)
+     * - executeAsync: (Closure, optional) This callable will
+     *   receive as parameter:
+     *   - $command: (CommandInterface, optional)
+     *
+     * @return S3Client
+     */
+    private function getS3ClientMock(
```

Comment:
All of this is hard to follow- Is there a reason `UsesServiceTrait` can't be used for mocking service clients?

Reply:
While `UsesServiceTrait` could be useful for some cases, there are others where I need to validate the parameters the command is populated with, based on the request being performed. For example [this](https://github.com/aws/aws-sdk-php/pull/3079/files#diff-3e448d40361b9c61b4c3e7b349cbd8dd6ac72d725de83e283298093bd5901161R548) test, and with this way to mock the client I can easily do it. So I dont see the need to remove this implementation in favor of `UsesServiceTrait` for now.